jina/serve/instrumentation/_aio_client.py

```diff
@@ -0,0 +1,319 @@
+# Copyright The OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import functools
+from collections import OrderedDict
+
+import grpc
+from grpc.aio import ClientCallDetails
+from opentelemetry import context
+from opentelemetry.instrumentation.grpc._client import (
+    OpenTelemetryClientInterceptor,
+    _carrier_setter,
+)
+from opentelemetry.instrumentation.utils import _SUPPRESS_INSTRUMENTATION_KEY
+from opentelemetry.propagate import inject
+from opentelemetry.semconv.trace import SpanAttributes
+from opentelemetry.trace.status import Status, StatusCode
+
+
+def _unary_done_callback(span, code, details):
+    def callback(call):
+        try:
+            span.set_attribute(
+                SpanAttributes.RPC_GRPC_STATUS_CODE,
+                code.value[0],
+            )
+            if code != grpc.StatusCode.OK:
+                span.set_status(
+                    Status(
+                        status_code=StatusCode.ERROR,
+                        description=details,
+                    )
+                )
+        finally:
+            span.end()
+
+    return callback
+
+
+class _BaseAioClientInterceptor(OpenTelemetryClientInterceptor):
```

Comment:
why do we need to copy this code here?

Reply:
The opentelemty-python package doesn't yet support `grpc.aio.*` implementations. The interfaces are same or similar to the normal grpc implementation but the `grpc.aio.*` expects async methods which has been provided in this code. This can be removed once the official package provides support.